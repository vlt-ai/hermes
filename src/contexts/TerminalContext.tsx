import * as React from "react"
export const TerminalContext = React.createContext(null)

export const TerminalContextProvider = (props: any) => {
  const { children } = props
  const [editorInput, setEditorInput] = React.useState("")
  const [bufferedContent, setBufferedContent] = React.useState(<></>)
  const [commandsHistory, setCommandsHistory] = React.useState([])
  const [historyPointer, setHistoryPointer] = React.useState(null)

  React.useEffect(() => {
    setHistoryPointer(commandsHistory.length);
  }, [commandsHistory]);

  const appendCommandToHistory = (command: string) => {
    if (!command) return
    setCommandsHistory(commandsHistory.concat(command));
  }

  const getPreviousCommand = () => {
    if (historyPointer === 0) {
      if (commandsHistory.length === 0) return ""
      return commandsHistory[0]
    }

    const command = commandsHistory[historyPointer - 1];
    if (historyPointer > 0) {
      setHistoryPointer(historyPointer - 1);
    }

    return command;
  };

  const getNextCommand = () => {
    if (historyPointer + 1 <= commandsHistory.length) {
      if (historyPointer === commandsHistory.length) {
        return ""
      }
      const command = commandsHistory[historyPointer + 1];
      setHistoryPointer(historyPointer + 1);
      return command;
    }
    return ""
  }

  const setNewline = (prompt: string) => setBufferedContent(bufferedContent => <>{bufferedContent}{prompt}</>)

  return (
    <TerminalContext.Provider
      value={{
        editorInput, 
        setEditorInput,
        bufferedContent,
        setBufferedContent,
        appendCommandToHistory,
        getPreviousCommand,
        getNextCommand
      }}
    >
      {children}
    </TerminalContext.Provider>
  );
};

export default {
  TerminalContext,
  TerminalContextProvider
};
