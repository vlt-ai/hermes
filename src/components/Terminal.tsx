import * as React from "react";
import PropTypes from "prop-types";
import useClippy from "use-clippy"
import { useWindowSize } from '@react-hook/window-size'
import { isMobile } from "react-device-detect";
import { Rnd } from 'react-rnd'

import { StyleContext } from "../contexts/StyleContext";
import { ThemeContext } from "../contexts/ThemeContext";
import { useClickOutsideEvent } from "../hooks/terminal";

import Controls from "./Controls";
import Editor from "./Editor";

export default function Terminal(props: any) {
  const wrapperRef = React.useRef(null);
  const [consoleFocused, setConsoleFocused] = React.useState(!isMobile);
  const style = React.useContext(StyleContext);
  const themeStyles = React.useContext(ThemeContext);

  useClickOutsideEvent(wrapperRef, consoleFocused, setConsoleFocused);

  // get all props destructively
  const {
    draggable,
    commandPassThrough,
    blockInput,
    isLoading,
    caret,
    theme,
    showControlButtons,
    controlButtonLabels,
    prompt,
    commands,
    welcomeMessage,
    errorMessage,
    enableInput
  } = props

  const rnd = React.createRef<Rnd>()
  const size = useWindowSize()

  const minimize = () => {
    if (rnd.current) rnd.current.updateSize({ width: 500, height: 200 })
  }

  const maximize = () => {
    const [width, height] = size
    if (rnd.current) rnd.current.updateSize({ width, height })
  }

  return (
    <>
      {
        draggable ?
          <Rnd
            ref={rnd}
            default={{
              x: 150,
              y: 205,
              width: 500,
              height: 200
            }
            }
            minWidth={500}
            minHeight={200}
            bounds="window"
            cancel=".cancel"
          >
            <div
              ref={wrapperRef}
              id={style.terminalContainer}
              className={`${style[`theme--${theme}`]}`}
            >
              <div className={`${style.terminal}`} style={{ background: themeStyles.themeToolbarColor, color: themeStyles.themeColor }}>
                <Controls
                  minimize={minimize}
                  maximize={maximize}
                  consoleFocused={consoleFocused}
                  showControlButtons={showControlButtons}
                  controlButtonLabels={controlButtonLabels}
                />
                <div className="cancel" style={{ width: "100%", height: "100%" }}>
                  <Editor
                    commandPassThrough={commandPassThrough}
                    blockInput={blockInput}
                    isLoading={isLoading}
                    caret={caret}
                    consoleFocused={consoleFocused}
                    prompt={prompt}
                    commands={commands}
                    welcomeMessage={welcomeMessage}
                    errorMessage={errorMessage}
                    enableInput={enableInput}
                  />
                </div>
              </div>
            </div>
          </Rnd > : <div
            ref={wrapperRef}
            id={style.terminalContainer}
            className={`${style[`theme--${theme}`]}`}
          >
            <div className={`${style.terminal}`} style={{ background: themeStyles.themeToolbarColor, color: themeStyles.themeColor }}>
              <Controls
                minimize={minimize}
                maximize={maximize}
                consoleFocused={consoleFocused}
                showControlButtons={showControlButtons}
                controlButtonLabels={controlButtonLabels}
              />
              <div className="cancel" style={{ width: "100%", height: "100%" }}>
                <Editor
                  commandPassThrough={commandPassThrough}
                  blockInput={blockInput}
                  isLoading={isLoading}
                  caret={caret}
                  consoleFocused={consoleFocused}
                  prompt={prompt}
                  commands={commands}
                  welcomeMessage={welcomeMessage}
                  errorMessage={errorMessage}
                  enableInput={enableInput}
                />
              </div>
            </div>
          </div>
      }
    </>
  );
}

Terminal.propTypes = {
  draggable: PropTypes.bool,
  commandPassThrough: PropTypes.func,
  blockInput: PropTypes.bool,
  isLoading: PropTypes.bool,
  enableInput: PropTypes.bool,  //added both props with their types
  caret: PropTypes.bool,
  theme: PropTypes.string,
  showControlButtons: PropTypes.bool,
  controlButtonLabels: PropTypes.arrayOf(PropTypes.string),
  prompt: PropTypes.string,
  commands: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.node
  ])),
  welcomeMessage: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.node]),
  errorMessage: PropTypes.string
};

Terminal.defaultProps = {
  draggable: false,
  commandPassThrough: null,
  blockInput: false,
  isLoading: false,
  enableInput: true,
  caret: true,
  theme: "light",
  showControlButtons: true,
  controlButtonLabels: ["close", "minimize", "maximize"],
  prompt: ">>>",
  commands: {},
  welcomeMessage: "",
  errorMessage: "not found!"
};
