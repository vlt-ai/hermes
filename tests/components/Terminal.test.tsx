import * as React from "react";
import renderer from "react-test-renderer";
import ContextProvider from "../../src/contexts";
import { TerminalContextProvider } from "../../src/contexts/TerminalContext";
import Terminal from "../../src/components/Terminal";

let props: any;
const renderWrapper = () => (
  <ContextProvider>
    <TerminalContextProvider>
      <Terminal {...props} />
    </TerminalContextProvider>
  </ContextProvider>
);

it("Terminal renders correctly", () => {
  expect(renderer.create(renderWrapper()).toJSON()).toMatchSnapshot();
});
