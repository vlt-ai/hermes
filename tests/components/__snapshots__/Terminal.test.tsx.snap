// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Terminal renders correctly 1`] = `
<div
  className="theme--light"
  id="terminalContainer"
>
  <div
    className="terminal"
    style={
      Object {
        "background": "#d8d8d8",
        "color": "#333333",
      }
    }
  >
    <div
      className="controls"
    >
      <div
        className="consoleCtrl close"
      />
      <div
        className="consoleCtrl minimize"
      />
      <div
        className="consoleCtrl maximize"
      />
    </div>
    <div
      className="editor"
      style={
        Object {
          "background": "#fdf6e4",
        }
      }
    >
      
      
      <span
        style={
          Object {
            "color": "#a917a8",
          }
        }
      >
        &gt;&gt;&gt;
      </span>
      <div
        className="lineText"
      >
        <span
          className="preWhiteSpace"
        >
          
        </span>
        <span
          className="caret"
        >
          <span
            className="caretAfter"
            style={
              Object {
                "background": "#333333",
              }
            }
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
